src/test/connect-test.c

```diff
@@ -0,0 +1,435 @@
+
+enum connection_status {
+        CONNECTION_UNDEFINED = 0,
+        CONNECTION_TIMEOUT,
+        CONNECTION_REFUSED,
+        CONNECTION_SUCCEEDED,
+};
+
+struct context {
+        enum connection_status ipv4;
+        enum connection_status ipv6;
+        enum connection_status timeout;
+};
+
+static const char *enum_to_s(enum connection_status status) {
+        switch (status) {
+        case CONNECTION_TIMEOUT:
+                return INPUT_CONNECTION_TIMEOUT;
+        case CONNECTION_REFUSED:
+                return INPUT_CONNECTION_REJECTED;
+        case CONNECTION_SUCCEEDED:
+                return INPUT_CONNECTION_SUCCESS;
+        default:
+                return "undefined";
+        }
+}
+
+static int timeout_handler(sd_event_source *event_source, usec_t usec, void *data) {
+        struct context *ctx = data;
+
+        ctx->timeout = CONNECTION_TIMEOUT;
+
+        sd_event_exit(sd_event_source_get_event(event_source), 0);
+        return 0;
+}
+
+static int server_handler(sd_event_source *event_source, int fd, uint32_t revents, void *data) {
+        int conn_fd, r;
+        union sockaddr_union addr, addr_name = {};
+        socklen_t addr_length = sizeof(addr_name), len;
+
+        r = getsockname(fd, &addr_name.sa, &addr_length);
+        if (r < 0) {
+                sd_event_exit(sd_event_source_get_event(event_source), r);
+                return -errno;
+        }
+
+        if (addr_name.sa.sa_family == AF_INET)
+                len = sizeof(addr.in);
+        else if (addr_name.sa.sa_family == AF_INET6)
+                len = sizeof(addr.in6);
+        else {
+                sd_event_exit(sd_event_source_get_event(event_source), -1);
+                return -EAFNOSUPPORT;
+        }
+
+        conn_fd = accept4(fd, &addr.sa, &len, SOCK_CLOEXEC);
+
+        if (conn_fd < 0) {
+                log_error_errno(errno, "Error accepting connection (fd: %d): %m", fd);
+                return -errno;
+        }
+
+        /*
+         * Only one connection expected -- return -1 so that the server handler
+         * is not called again.
+         */
+        return -1;
+}
+
+static int server_handler_4(sd_event_source *event_source, int fd, uint32_t revents, void *data) {
+        return server_handler(event_source, fd, revents, data);
+}
+
+static int server_handler_6(sd_event_source *event_source, int fd, uint32_t revents, void *data) {
+        return server_handler(event_source, fd, revents, data);
+}
+
+static int conn_handler(sd_event_source *event_source, int fd, uint32_t revents, void *data) {
+        int error, r;
+        struct context *ctx = data;
+        unsigned int length = sizeof(error);
+        struct sockaddr addr = { 0 };
+        socklen_t addr_length = sizeof(addr);
+        enum connection_status *status;
+
+        r = getsockname(fd, &addr, &addr_length);
+        if (r < 0) {
+                sd_event_exit(sd_event_source_get_event(event_source), r);
+                return r;
```

Comment:
this doesn't look right... please don't make up error numbers. instead propagate the right `errno` values, i.e. `return -errno` here